src/Equinox.MessageDb/MessageDb.fs

```diff
@@ -322,63 +350,94 @@ type MessageDbContext(connection : MessageDbConnection, batchOptions : BatchOpti
         | GatewaySyncResult.ConflictUnknown ->
             return SyncResult.Conflict  (fun ct -> x.Reload(fold, state, streamName, (*requireLeader*)true, token, log) |> Async.startAsTask ct)
         | GatewaySyncResult.Written token' ->
-            return SyncResult.Written   (token', fold state (Seq.ofArray events)) }
+            let state' = fold state (Seq.ofArray events)
+            match access with
+            | None | Some AccessStrategy.LatestKnownEvent -> ()
+            | Some (AccessStrategy.RollingSnapshots(_, toSnap)) ->
+              if Token.shouldSnapshot context.BatchOptions.BatchSize token token' then
+                do! x.StoreSnapshot(categoryName, streamId, log, ctx, token', toSnap state')
+            return SyncResult.Written   (token', state') }
 
 type private Folder<'event, 'state, 'context>(category : Category<'event, 'state, 'context>, fold : 'state -> 'event seq -> 'state, initial : 'state, ?readCache) =
-    let batched log streamName requireLeader ct = category.Load(fold, initial, streamName, requireLeader, log) |> Async.startAsTask ct
+    let batched log categoryName streamId streamName requireLeader ct = category.Load(fold, initial, categoryName, streamId, streamName, requireLeader, log) |> Async.startAsTask ct
+
     interface ICategory<'event, 'state, 'context> with
-        member _.Load(log, _categoryName, _streamId, streamName, allowStale, requireLeader, ct) = task {
+        member _.Load(log, categoryName, streamId, streamName, allowStale, requireLeader, ct) = task {
             let act = Activity.Current
             match readCache with
-            | None -> return! batched log streamName requireLeader ct
+            | None -> return! batched log categoryName streamId streamName requireLeader ct
             | Some (cache : ICache, prefix : string) ->
                 let! cacheItem = cache.TryGet(prefix + streamName)
                 if act <> null then act.AddCacheHit(match cacheItem with ValueNone -> false | _ -> true) |> ignore
                 match cacheItem with
-                | ValueNone -> return! batched log streamName requireLeader ct
+                | ValueNone -> return! batched log categoryName streamId streamName requireLeader ct
                 | ValueSome tokenAndState when allowStale -> return tokenAndState
                 | ValueSome (token, state) -> return! category.Reload(fold, state, streamName, requireLeader, token, log) }
-        member _.TrySync(log, _categoryName, _streamId, streamName, context, _init, token, originState, events, _ct) = task {
-            match! category.TrySync(log, fold, streamName, token, originState, events, context) with
+
+        member _.TrySync(log, categoryName, streamId, streamName, context, _init, token, originState, events, ct) = task {
+            let! result = category.TrySync(log, fold, categoryName, streamId, streamName, token, originState, events, context) |> Async.startAsTask ct
+            match result with
```

Comment:
why you no like `match!` (there were even opportunities missed above!)

Reply:
Got tired of explaining "It's just a let! binding followed by a match expression" 😂.

I'll change